### PR TITLE
Update AnyVolumeDataSource feature gate to beta

### LINKED ISSUE #99955: Update AnyVolumeDataSource feature gate to beta
#### What would you like to be added:

Update AnyVolumeDataSource feature gate to beta

#### Why is this needed:

KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1495-volume-populators/README.md

### PR DESCRIPTION
**Pull request by bswartz**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Default AnyVolumeDataSource feature gate to enabled, as the feature is moving to beta.

#### Which issue(s) this PR fixes:
Fixes #99955

#### Does this PR introduce a user-facing change?
```release-note
The AnyVolumeDataSource feature is now beta, and the feature gate is enabled by default. In order to provide user feedback on PVCs with data sources, deployers must install the VolumePopulators CRD and the data-source-validator controller.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
[KEP]: 1495
```

### CODE REVIEW

**Comment by xing-yang** [comment]
This needs to be updated to "beta" as well:
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.4/pkg/apis/core/types.go#L475

**Comment by xing-yang** [comment]
@bswartz Can you also add that this feature is Beta in the release note.

**Comment by bswartz** [comment]
> @bswartz Can you also add that this feature is Beta in the release note.

Done

**Comment by liggitt** [comment]
from the KEP (https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#alpha---beta-graduation)

* Before going to beta, we need a clear notion of how data populators should work.
* We will need a simple and lightweight implementation of a data populator that can be used by the E2E test suite to exercise the functionality.
* Automated tests that create a volume from a data source handled by a populator, to validate that the data source functionality works, and that any other required functionality for data population is working.
* We will need to see several implementations of working data populators that solve real world problems implemented in the community.
* volume-data-source-validator and associated VolumePopulator CRD should be released with beta versions available.

Can you give links to those either here or in the KEP or in the enhancements issue to verify alpha → beta graduation criteria are met?

**Comment by liggitt** [comment]
(mechanical changes in this PR lgtm)

**Comment by bswartz** [comment]
> from the KEP (https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#alpha---beta-graduation)
> 
>     * Before going to beta, we need a clear notion of how data populators should work.

There is a library that does most of the work for this. https://github.com/kubernetes-csi/lib-volume-populator

>     * We will need a simple and lightweight implementation of a data populator that can be used by the E2E test suite to exercise the functionality.

The example populator included with the above library is called "hello-populator" and provides this functionality. https://github.com/kubernetes-csi/lib-volume-populator/tree/master/example/hello-populator

>     * Automated tests that create a volume from a data source handled by a populator, to validate that the data source functionality works, and that any other required functionality for data population is working.

The test exists and will soon be merged. https://github.com/kubernetes/kubernetes/pull/108664

>     * We will need to see several implementations of working data populators that solve real world problems implemented in the community.

I don't have a comprehensive list of these, but here is one:
https://openebs.io/blog/a-hands-on-demo-of-volume-populator-using-openebs-lvm-csi-driver

Other projects in the works include:
* A NetApp implementation of foriegn-volume imports based on data populators
* Restoring from backups using a data populator

@xing-yang  might have more examples here

>     * volume-data-source-validator and associated VolumePopulator CRD should be released with beta versions available.

This is here: https://github.com/kubernetes-csi/volume-data-source-validator/blob/release-1.0/client/config/crd/populator.storage.k8s.io_volumepopulators.yaml

> Can you give links to those either here or in the KEP or in the enhancements issue to verify alpha → beta graduation criteria are met?

I hope I covered all of these above. Let me know if you need more details or and other supplemental material.

**Comment by xing-yang** [comment]
Other than the implementation examples listed by @bswartz, we also have plan to do a POC with Volume Populator.

**Comment by liggitt** [comment]
> There is a library that does most of the work for this. https://github.com/kubernetes-csi/lib-volume-populator
> 
> > ```
> > * We will need a simple and lightweight implementation of a data populator that can be used by the E2E test suite to exercise the functionality.
> > ```

Thanks, that's helpful. As an aside, I see it got tagged as v1.0.0 built to use the beta APIs. In general (in the future), we should make it clear when components have hard requirements on non-GA APIs and avoid marking them as v1.0.0.

**Comment by liggitt** [comment]
@thockin, it looks like this is using a similar decorate-on-read approach to Service.

Is a similar change to https://github.com/kubernetes/kubernetes/pull/107847/files needed here?

**Comment by liggitt** [comment]
API comment, feature gate change, and KEP graduation references lgtm. Routing to thockin for the question on default-on-read behavior

**Comment by bswartz** [comment]
> Thanks, that's helpful. As an aside, I see it got tagged as v1.0.0 built to use the beta APIs. In general (in the future), we should make it clear when components have hard requirements on non-GA APIs and avoid marking them as v1.0.0.

Yes it's complicated to evolve a partially in-tree/partially out-of-tree feature through alpha->beta->GA. What version number would have been better to use for the out-of-tree beta stuff? I thought that anything starting with 0.x had bad implications for semver in golang, so 1.0 was the obvious choice (especially since there were no breaking changes made).

**Comment by bswartz** [comment]
> API comment, feature gate change, and KEP graduation references lgtm. Routing to thockin for the question on default-on-read behavior

I saw an earlier reference to an issue with default-on-read for Services, and I wasn't able to see how the dataSourceRef field could have the same problems, but I'd be grateful to have another set of eyes on it.

**Comment by liggitt** [comment]
tagging it v0 until it is built on top of stable APIs would make more sense to me... my understanding is that anything built on top of the current v1.0.0 library will break when the CRD promotes to v1 and stops serving v1beta1, right? if so, that doesn't match what I would expect from a v1 library

**Comment by liggitt** [comment]
> > API comment, feature gate change, and KEP graduation references lgtm. Routing to thockin for the question on default-on-read behavior
> 
> I saw an earlier reference to an issue with default-on-read for Services, and I wasn't able to see how the dataSourceRef field could have the same problems, but I'd be grateful to have another set of eyes on it.

tl;dr is that the decorator doesn't run on the old object read from etcd when doing an update, so if your PrepareForUpdate or update validation would be unhappy with the new object being defaulted and the old object not being defaulted, you'll need to default the old object explicitly in those paths

**Comment by bswartz** [comment]
> tl;dr is that the decorator doesn't run on the old object read from etcd when doing an update, so if your PrepareForUpdate or update validation would be unhappy with the new object being defaulted and the old object not being defaulted, you'll need to default the old object explicitly in those paths

Is there an easy workaround that I can copy? I'll take another look at how this was fixed for the Service object to see if I missed something.

**Comment by liggitt** [comment]
it looks like persistentvolumeclaimStrategy#PrepareForUpdate is already calling the same default-on-read normalization as defaultOnReadPvc on the new PVC:

https://github.com/kubernetes/kubernetes/blob/e0ca5cfd73bd046ee5fbdef50e073842e6fbc52f/pkg/registry/core/persistentvolumeclaim/strategy.go#L99-L114

But it's not touching the old PVC. I don't know if validation is counting on the old PVC having been normalized by the default-on-read path or not, but it is not currently being normalized.

**Comment by bswartz** [comment]
> it looks like persistentvolumeclaimStrategy#PrepareForUpdate is already calling the same default-on-read normalization as defaultOnReadPvc on the new PVC:
> But it's not touching the old PVC. I don't know if validation is counting on the old PVC having been normalized by the default-on-read path or not, but it is not currently being normalized.

I understood that that function treated the "oldPvc" as a read-only object and that newPvc was the one to be updated. How could anyone be interested in the contents of the oldPvc object after calling PrepareForUpdate? Maybe I'm not familiar enough the the API server guts, but I don't see any benefit to also fixing the oldPvc in that function (if it isn't already accurate)

**Comment by liggitt** [comment]
Some validation disallows specific transitions or enforces immutable fields based on comparing the old and new objects. If the new object had migration applied and the old object didn't, the user could get an error saying a field was not allowed to be changed in some way that would make no sense to them (since from their perspective they weren't changing the field)

**Comment by bswartz** [comment]
I see what you're saying. You're not worried about bogus output from the function but the possibility that a bogus input could cause it to misbehave. I'll look for that specifically.

**Comment by bswartz** [comment]
Okay I re-reviewed the code and this is a case that we considered extensively when we first designed this (with a huge amount of help from @thockin). If you follow the PrepareForUpdate() function logic, there are several function calls but they're all very small. There's exactly one place where the contents of the old PVC is considered, and it's written to look at **both** the old field and the new field, such that it makes the correct decision whether the old object is upgraded or not.

https://github.com/kubernetes/kubernetes/blob/e0ca5cfd73bd046ee5fbdef50e073842e6fbc52f/pkg/api/persistentvolumeclaim/util.go#L92

The huge comment for EnforceDataSourceBackwardsCompatibility() tries to explain exactly what's going on here so that future versions of ourselves can remember why this code is the way it is. I'm confident there are no issues if someone tries to update an un-upgraded object.

**Comment by liggitt** [comment]
The last commit in https://github.com/liggitt/kubernetes/commits/any-volume-data-source-beta adds an integration test that demonstrates the problems if someone tries to update a non-upgraded PVC object. All three update methods fail with the feature enabled:
1. write original content that just sets `dataSource`
2. send a no-op patch of `{}` that lets the server compute the updated object internally based on the persisted object
3. get from the server and put the content without changing it

All three update methods work with the AnyVolumeDataSource feature gate disabled.

```
        upgrade_test.go:90: PVC stored in etcd {"kind":"PersistentVolumeClaim","apiVersion":"v1","metadata":{"name":"test-old-pvc","namespace":"old-pvc-ns","uid":"08675309-9376-9376-9376-086753099999","creationTimestamp":"2022-03-29T13:36:09Z"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"10G"}},"dataSource":{"apiGroup":null,"kind":"PersistentVolumeClaim","name":"foo"}},"status":{}}
        upgrade_test.go:96: write of original content failed: PersistentVolumeClaim "test-old-pvc" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims
              core.PersistentVolumeClaimSpec{
              	... // 5 identical fields
              	VolumeMode:    &"Filesystem",
              	DataSource:    &{Kind: "PersistentVolumeClaim", Name: "foo"},
            - 	DataSourceRef: nil,
            + 	DataSourceRef: &core.TypedLocalObjectReference{Kind: "PersistentVolumeClaim", Name: "foo"},
              }
        upgrade_test.go:104: no-op patch failed: PersistentVolumeClaim "test-old-pvc" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims
              core.PersistentVolumeClaimSpec{
              	... // 5 identical fields
              	VolumeMode:    &"Filesystem",
              	DataSource:    &{Kind: "PersistentVolumeClaim", Name: "foo"},
            - 	DataSourceRef: nil,
            + 	DataSourceRef: &core.TypedLocalObjectReference{Kind: "PersistentVolumeClaim", Name: "foo"},
              }
        upgrade_test.go:116: no-op get/put failed: PersistentVolumeClaim "test-old-pvc" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims
              core.PersistentVolumeClaimSpec{
              	... // 5 identical fields
              	VolumeMode:    &"Filesystem",
              	DataSource:    &{Kind: "PersistentVolumeClaim", Name: "foo"},
            - 	DataSourceRef: nil,
            + 	DataSourceRef: &core.TypedLocalObjectReference{Kind: "PersistentVolumeClaim", Name: "foo"},
              }
```

**Review by thockin — COMMENTED** [review]
This seems like the right fix.  It's an unfortunate wart, but <shrug>.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve
